skbio/tree/_tree.py

```diff
@@ -3098,3 +3098,33 @@ def shuffle(self, k=None, names=None, shuffle_f=np.random.shuffle, n=1):
 
             yield self
             counter += 1
+
+    def _cache_leafcount(self):
```

Comment:
These caches need to be invalidated on topology changes. Please see [`TreeNode.invalidate_caches`](https://github.com/biocore/scikit-bio/blob/master/skbio/tree/_tree.py#L1364).